From 54d72b51a4a64212bb8cc1e23f666dafabe49931 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Tue, 23 Mar 2021 17:48:37 +0100
Subject: [PATCH 1/1] efi_loader: update system table in SetVirtualAddressMap

In SetVirtualAddressMap() we must update the following fields of the system
table:

* Hdr.CRC32
* FirmwareVendor
* ConfigurationTable
* RuntimeServices

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_boottime.c | 4 ++--
 lib/efi_loader/efi_runtime.c  | 9 +++++++++
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/lib/efi_loader/efi_boottime.c b/lib/efi_loader/efi_boottime.c
index 41b8949b04..607f1453c8 100644
--- a/lib/efi_loader/efi_boottime.c
+++ b/lib/efi_loader/efi_boottime.c
@@ -3759,9 +3759,7 @@ struct efi_system_table __efi_runtime_data systab = {
 		.revision = EFI_SPECIFICATION_VERSION,
 		.headersize = sizeof(struct efi_system_table),
 	},
-	.fw_vendor = firmware_vendor,
 	.fw_revision = FW_VERSION << 16 | FW_PATCHLEVEL << 8,
-	.runtime = &efi_runtime_services,
 	.nr_tables = 0,
 	.tables = NULL,
 };
@@ -3789,6 +3787,8 @@ efi_status_t efi_initialize_system_table(void)
 	systab.con_out = &efi_con_out;
 	systab.std_err = &efi_con_out;
 	systab.boottime = &efi_boot_services;
+	systab.fw_vendor = firmware_vendor,
+	systab.runtime = &efi_runtime_services,
 
 	/* Set CRC32 field in table headers */
 	efi_update_table_header_crc32(&systab.hdr);
diff --git a/lib/efi_loader/efi_runtime.c b/lib/efi_loader/efi_runtime.c
index 93a695fc27..8fa8909a15 100644
--- a/lib/efi_loader/efi_runtime.c
+++ b/lib/efi_loader/efi_runtime.c
@@ -35,6 +35,9 @@ struct efi_runtime_mmio_list {
 LIST_HEAD(efi_runtime_mmio);
 
 static efi_status_t __efi_runtime EFIAPI efi_unimplemented(void);
+static __efi_runtime
+efi_status_t EFIAPI efi_convert_pointer_runtime(efi_uintn_t debug_disposition,
+						void **address);
 
 /*
  * TODO(sjg@chromium.org): These defines and structures should come from the ELF
@@ -661,8 +664,14 @@ static __efi_runtime void efi_relocate_runtime_table(ulong offset)
 	 * EFI_RT_SUPPORTED_CONVERT_POINTER as required by the UEFI spec 2.8.
 	 */
 
+	/* Update system table */
+	efi_convert_pointer(0, &systab.fw_vendor);
+	efi_convert_pointer(0, &systab.runtime);
+	efi_convert_pointer(0, &systab.tables);
+
 	/* Update CRC32 */
 	efi_update_table_header_crc32(&efi_runtime_services.hdr);
+	efi_update_table_header_crc32(&systab.hdr);
 }
 
 /* Relocate EFI runtime to uboot_reloc_base = offset */
-- 
2.31.0

